### PR TITLE
CMake framework support

### PR DESCRIPTION
**Pull request by simoncent**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -178,6 +178,18 @@ jobs:
             config: Release,
             args: -DBUILD_SHARED_LIBS=ON
           }
+        - {
+            name: "MacOSX/Dll/X64/Debug/Framework",
```
Comment (simoncent) [inline .github/workflows/build.yml:182]: I was unsure of how you would like this named. The frameworks are still shared libraries, so I thought I would leave it under the `Dll` convention, happy to change it though.

Comment (xnorpx) [reply .github/workflows/build.yml:182]: Replace Dll with Framework i think is fine

File: CMakeLists.txt [hunk 2]
```diff
@@ -179,6 +175,9 @@ set(Opus_PUBLIC_HEADER
     ${CMAKE_CURRENT_SOURCE_DIR}/include/opus_projection.h
     ${CMAKE_CURRENT_SOURCE_DIR}/include/opus_types.h)
 
+add_library(opus ${opus_sources} ${opus_sources_float} ${Opus_PUBLIC_HEADER})
```
Comment (xnorpx) [inline CMakeLists.txt:232]: Remove ${Opus_PUBLIC_HEADER} for now
If you want then you can create another pull request were you add all headers including the public header to the project.

File: CMakeLists.txt [hunk 3]
```diff
@@ -37,6 +37,10 @@ option(OPUS_INSTALL_PKG_CONFIG_MODULE "Install PkgConfig module" ON)
 option(OPUS_INSTALL_CMAKE_CONFIG_MODULE "Install CMake package config module"
        ON)
 
+if(APPLE)
+  option(BUILD_FRAMEWORK "Build Framework bundle for Apple systems" OFF)
```
Comment (xnorpx) [inline CMakeLists.txt:41]: let's use OPUS_BUILD_FRAMEWORK to keep it OPUS specific

Comment (xnorpx) [inline CMakeLists.txt:41]: add the option below were all the other options are specified so it is printed out into the help

File: CMakeLists.txt [hunk 4]
```diff
@@ -217,10 +221,6 @@ endif()
 
 feature_summary(WHAT ALL)
 
-add_library(opus ${opus_sources} ${opus_sources_float})
```
Comment (xnorpx) [inline CMakeLists.txt:220]: revert this change for now

**Comment by xnorpx** [comment]
Squash the change into 2 commits, one for the build file change and one for the cmakefile change

File: CMakeLists.txt (hunk 3 above)
Comment (simoncent) [reply CMakeLists.txt:41]: `BUILD_FRAMEWORK` is a CMake variable, like `BUILD_SHARED_LIBS`, it's pretty normal to add them as options like this. I thought it a bit odd that `BUILD_SHARED_LIBS` is prefixed with `OPUS_` to be honest. The Ogg CMake build doesn't prefix them.

I can change it for consistency, but it does seem odd to me to duplicate them in this way, IMO it would be better to remove the `OPUS_` prefix where providing an option to set a CMake variable.

File: .github/workflows/build.yml (hunk 1 above)
Comment (simoncent) [reply .github/workflows/build.yml:182]: OK, sounds good.

**Comment by xnorpx** [comment]
Once the majority of updates are done we are going to run cmake format on the file so don't bother so much about the style.

**Comment by simoncent** [comment]
> Squash the change into 2 commits, one for the build file change and one for the cmakefile change

Wouldn't it be cleaner to do a sqaush merge when merging the PR after approval, that is what I was expecting to do?

File: CMakeLists.txt (hunk 3 above)
Comment (xnorpx) [reply CMakeLists.txt:41]: I see, then please duplicate similar to BUILD_SHARED_LIBS. I want people to be able to turn on and off features without affecting other projects if they are build through fetchcontent etc.

**Comment by xnorpx** [comment]
Please ensure there is 2 different commit, as the build changes will stay in this repo while the cmakechanges will be sent as a patch to the xiph repo. They haven't agreed on adding cmake build to their CI so I need to keep it here.

**Comment by simoncent** [comment]
OK, I will do that then.

File: CMakeLists.txt (hunk 2 above)
Comment (simoncent) [reply CMakeLists.txt:232]: The public headers have to be included here or they are not installed into the built framework.

File: CMakeLists.txt (hunk 4 above)
Comment (simoncent) [reply CMakeLists.txt:220]: The public headers have to be included here or they are not installed into the built framework.

File: CMakeLists.txt (hunk 3 above)
Comment (simoncent) [reply CMakeLists.txt:41]: I haven't used `FetchContent`, but with `ExternalProject_Add` you still need to explicitly pass the settings of those variables, so adding them as options directly without the prefix wouldn't interfere with other projects. The CMake documentation even states they often added directly as options: https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

File: CMakeLists.txt (hunk 4 above)
Comment (xnorpx) [reply CMakeLists.txt:220]: I see, then leave it and I add the other headers at a later stage.

Comment (simoncent) [reply CMakeLists.txt:220]: Thanks!

File: CMakeLists.txt (hunk 3 above)
Comment (simoncent) [reply CMakeLists.txt:41]: > `BUILD_FRAMEWORK` is a CMake variable,

Sorry, I was wrong here, I was getting mixed up with a different variable! `BUILD_FRAMEWORK` is not a CMake variable like `BUILD_SHARED_LIBS`.

**Comment by simoncent** [comment]
I've made all the requested changes now.

**Comment by xnorpx** [comment]
Note @simoncent I will deliver this one in the next cmake drop. So it will sit here for until the previous cmake drop has made it to xiph/opus master

**Comment by simoncent** [comment]
OK, that's fine, I'm using my fork in our builds anyway. Thanks for the help.
